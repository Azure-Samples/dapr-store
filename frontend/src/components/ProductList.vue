<template>
  <h1>PRODUCT LIST {{ viewType }}</h1>
</template>

<script>
export default {
  name: 'ProductList',

  props: {
    viewType: {
      type: String,
      required: true
    }
  }
}
</script>

<style scoped>

</style>