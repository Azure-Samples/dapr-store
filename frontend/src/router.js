import VueRouter from 'vue-router'

import ProductCatalog from './components/ProductCatalog'
import ProductOffers from './components/ProductOffers'
import Home from './components/Home'

const router = new VueRouter({
  routes: [
    {
      path: '/catalog',
      component: ProductCatalog
    },
    {
      path: '/offers',
      component: ProductOffers
    },
    {
      path: '/',
      component: Home
    }
  ]
})

export default router