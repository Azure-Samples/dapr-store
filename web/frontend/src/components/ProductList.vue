<!--
// ----------------------------------------------------------------------------
// Copyright (c) Ben Coleman, 2020
// Licensed under the MIT License.
//
// Dapr Store frontend - Reusable component that lists products
// ----------------------------------------------------------------------------
-->

<template>
  <div>
    <div v-if="!products" class="text-center">
      <b-spinner variant="success" style="width: 5rem; height: 5rem;" />
    </div>

    <b-card v-for="product in products" :key="product.id">
      <b-row>
        <b-col>
          <b-link :to="`/product/`+product.id">
            <b-card-title>
              {{ product.name }}
            </b-card-title>
          </b-link>
          <b-card-text>
            {{ product.description }}
            <br><br>
            <h4>Â£{{ product.cost }}</h4>
          </b-card-text>

          <b-button :disabled="!isLoggedIn()" href="#" variant="primary" class="d-none d-md-inline" @click="addToCart(product)">
            <fa icon="shopping-cart" />
            &nbsp; Add to Cart
          </b-button>
        </b-col>

        <b-col class="flex-grow-0 d-none d-md-block">
          <div class="product-img">
            <span v-if="product.onOffer" class="onsale">On Sale</span>
            <b-link :to="`/product/`+product.id">
              <img :src="product.image">
            </b-link>
          </div>
        </b-col>
      </b-row>
    </b-card>
  </div>
</template>

<script>
import api from '../services/api'
import auth from '../services/auth'

export default {
  name: 'ProductList',


  props: {
    products: {
      type: Array,
      default: () => []
    }
  },

  methods: {
    async addToCart(product) {
      try {
        if (!auth.user()) { return }

        await api.cartAddAmount(auth.user().userName, product.id, +1)
        this.showToast('Added to your cart!', 'success', product)
      } catch (err) {
        this.showToast('Error adding to cart ðŸ˜« '+err.toString(), 'danger', product)
      }
    },

    isLoggedIn() {
      if (auth.user()) { return true }
      return false
    },

    showToast(msg, variant, product) {
      this.$bvToast.toast(`${product.name}`, {
        title: msg,
        variant: variant,
        autoHideDelay: 3000,
        appendToast: true,
        toaster: 'b-toaster-top-center',
        solid: true
      })
    }
  }
}
</script>

<style scoped>
.card {
  margin: 1rem;
}
.card-title {
  font-size: 2rem;
}
.product-img {
  float: right;
}
.product-img img {
  width: 12rem;
  border-radius: 0.5rem;
}
.card-text {
  font-size: 130%
}
.onsale {
  display: inline-block;
  position: absolute;
  bottom: 0;
  width: 12rem;
  height: 1.5rem;
  font-size: 1rem;
  color: white;
  background-color: rgb(185, 17, 17);
  text-align: center;
  border-bottom-left-radius: 0.5rem;
  border-bottom-right-radius: 0.5rem;
}
</style>