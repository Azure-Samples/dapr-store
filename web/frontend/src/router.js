// ----------------------------------------------------------------------------
// Copyright (c) Ben Coleman, 2020
// Licensed under the MIT License.
//
// Dapr Store frontend - client side routes
// ----------------------------------------------------------------------------

import { createRouter, createWebHistory } from 'vue-router'

import auth from './services/auth'

import ProductCatalog from './views/ProductCatalog.vue'
import ProductOffers from './views/ProductOffers.vue'
import ProductSearch from './views/ProductSearch.vue'
import ProductSingle from './views/ProductSingle.vue'
import ViewOrder from './views/ViewOrder.vue'
import Login from './views/Login.vue'
import Account from './views/Account.vue'
import Home from './views/Home.vue'
import About from './views/About.vue'
import Cart from './views/Cart.vue'

const router = createRouter({
  history: createWebHistory(),
  routes: [
    {
      name: 'home',
      path: '/',
      component: Home
    },
    {
      name: 'login',
      path: '/login',
      component: Login
    },
    {
      name: 'account',
      path: '/account',
      component: Account,
      beforeEnter: signedInCheck
    },
    {
      name: 'cart',
      path: '/cart',
      component: Cart,
      beforeEnter: signedInCheck
    },
    {
      name: 'about',
      path: '/about',
      component: About
    },
    {
      name: 'search',
      path: '/search/:query',
      component: ProductSearch
    },
    {
      name: 'catalog',
      path: '/catalog',
      component: ProductCatalog
    },
    {
      name: 'offers',
      path: '/offers',
      component: ProductOffers
    },
    {
      name: 'single-product',
      path: '/product/:id',
      component: ProductSingle
    },
    {
      name: 'view-order',
      path: '/order/:id',
      component: ViewOrder
      //beforeEnter: signedInCheck
    }
  ]
})

function signedInCheck(to, from, next) {
  const user = auth.user()
  if (!user || !user.username) {
    next('/login')
  } else {
    next()
  }
}

export default router
