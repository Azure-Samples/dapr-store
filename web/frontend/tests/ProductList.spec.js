import { vi, describe, it, expect } from 'vitest'
import { mount, flushPromises } from '@vue/test-utils'
import router from '@/router'

import ProductList from '@/components/ProductList.vue'

vi.mock('@/services/api')

// Load mock data
const mockJson = require('fs').readFileSync(__dirname + '/../../../testing/mock-data/products.json')
const mockProducts = JSON.parse(mockJson)

describe('ProductList.vue', () => {
  it('renders product list', async () => {
    const wrapper = mount(ProductList, {
      propsData: {
        products: mockProducts
      },
      global: {
        plugins: [router]
      }
    })

    await flushPromises()
    expect(wrapper.html()).toMatchSnapshot()
  })
})
