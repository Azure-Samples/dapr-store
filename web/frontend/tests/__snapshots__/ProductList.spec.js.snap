// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductList.vue > renders product list 1`] = `
"<div data-v-32d040e6="">
  <!--v-if-->
  <!--v-if-->
  <div data-v-32d040e6="" class="card">
    <div data-v-32d040e6="" class="card-body">
      <div data-v-32d040e6="" class="row">
        <div data-v-32d040e6="" class="col"><a data-v-32d040e6="" href="/product/prd1" class="">
            <div data-v-32d040e6="" class="card-title">Top Hat (6″)</div>
          </a>
          <div data-v-32d040e6="" class="card-text">Made from 100% wool and quite nice <h4 data-v-32d040e6="" class="mt-4">£39.95</h4>
          </div><button data-v-32d040e6="" disabled="" href="#" class="btn btn-primary d-none d-md-inline"><i data-v-32d040e6="" class="fa-solid fa-basket-shopping"></i> &nbsp; Add to Cart </button>
        </div>
        <div data-v-32d040e6="" class="col flex-grow-0 d-none d-md-block">
          <div data-v-32d040e6="" class="product-img">
            <!--v-if--><a data-v-32d040e6="" href="/product/prd1" class=""><img data-v-32d040e6="" src="/img/catalog/1.jpg"></a>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div data-v-32d040e6="" class="card">
    <div data-v-32d040e6="" class="card-body">
      <div data-v-32d040e6="" class="row">
        <div data-v-32d040e6="" class="col"><a data-v-32d040e6="" href="/product/prd2" class="">
            <div data-v-32d040e6="" class="card-title">Black &amp; Gold Tie Set</div>
          </a>
          <div data-v-32d040e6="" class="card-text">100% hand made silk includes tie, pocket square and cufflinks <h4 data-v-32d040e6="" class="mt-4">£18</h4>
          </div><button data-v-32d040e6="" disabled="" href="#" class="btn btn-primary d-none d-md-inline"><i data-v-32d040e6="" class="fa-solid fa-basket-shopping"></i> &nbsp; Add to Cart </button>
        </div>
        <div data-v-32d040e6="" class="col flex-grow-0 d-none d-md-block">
          <div data-v-32d040e6="" class="product-img"><span data-v-32d040e6="" class="onsale">On Sale</span><a data-v-32d040e6="" href="/product/prd2" class=""><img data-v-32d040e6="" src="/img/catalog/2.jpg"></a></div>
        </div>
      </div>
    </div>
  </div>
  <div data-v-32d040e6="" class="card">
    <div data-v-32d040e6="" class="card-body">
      <div data-v-32d040e6="" class="row">
        <div data-v-32d040e6="" class="col"><a data-v-32d040e6="" href="/product/prd3" class="">
            <div data-v-32d040e6="" class="card-title">Paisley Cravat Ascot Tie</div>
          </a>
          <div data-v-32d040e6="" class="card-text">Black &amp; Red Paisley Cravat <h4 data-v-32d040e6="" class="mt-4">£11.2</h4>
          </div><button data-v-32d040e6="" disabled="" href="#" class="btn btn-primary d-none d-md-inline"><i data-v-32d040e6="" class="fa-solid fa-basket-shopping"></i> &nbsp; Add to Cart </button>
        </div>
        <div data-v-32d040e6="" class="col flex-grow-0 d-none d-md-block">
          <div data-v-32d040e6="" class="product-img">
            <!--v-if--><a data-v-32d040e6="" href="/product/prd3" class=""><img data-v-32d040e6="" src="/img/catalog/3.jpg"></a>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div data-v-32d040e6="" class="position-fixed top-0 start-50 translate-middle-x p-3" style="z-index: 11;">
    <div data-v-32d040e6="" class="toast hide" role="alert">
      <div data-v-32d040e6="" class="d-flex">
        <div data-v-32d040e6="" class="toast-body fs-4"> was added to your cart!</div><button data-v-32d040e6="" type="button" class="btn-close me-2 m-auto" data-bs-dismiss="toast" aria-label="Close"></button>
      </div>
    </div>
  </div>
</div>"
`;
